app/models/organization/CreditTransaction.scala

```diff
@@ -0,0 +1,364 @@
+    for {
+      // Query: Sums up all spent credits since the transaction which are completed and subtracts refunded transactions.
+      freeCreditsAvailableResult <- q"""
+      SELECT COALESCE(SUM(credit_change), 0)
+      FROM webknossos.organization_credit_transactions
+      WHERE _organization = ${transaction._organization}
+        AND created_at >= ${transaction.createdAt}
+        AND (
+		      credit_change < 0
+          OR (credit_change > 0 AND refunded_transaction_id IS NOT NULL) -- Counts also revoked transactions
+          OR (credit_change > 0 AND expiration_date <= NOW()) -- Counts also expired transactions
+	      )
+		""".as[BigDecimal]
+      freeCreditsAvailable = freeCreditsAvailableResult.headOption.getOrElse(BigDecimal(0))
+
+      _ <- if (freeCreditsAvailable <= 0) {
+        // Fully spent, update state to 'Spent'
+        q"""
+        UPDATE webknossos.organization_credit_transactions
+        SET state = ${CreditTransactionState.Spent}, updated_at = NOW()
+        WHERE _id = ${transaction._id}
+      """.asUpdate
+      } else {
+        val stateOfExpiredTransaction = if (freeCreditsAvailable == transaction.creditChange) {
+          CreditTransactionState.Revoked
+        } else { CreditTransactionState.PartiallyRevoked }
+        val revokingTransaction = CreditTransaction(
+          ObjectId.generate,
+          transaction._organization,
+          -freeCreditsAvailable,
+          0,
+          None,
+          None,
+          s"Revoked expired credits for transaction ${transaction._id}",
+          None,
+          CreditTransactionState.Completed,
+          None,
+        )
+        for {
+          _ <- q"""
+            UPDATE webknossos.organization_credit_transactions
+            SET state = $stateOfExpiredTransaction, updated_at = NOW()
+            WHERE _id = ${transaction._id}
+          """.asUpdate
+          _ <- insertRevokingTransaction(revokingTransaction)
+        } yield ()
+      }
+      _ = logger.info(s"revokeExpiredCreditsTransaction for transaction ${transaction._id} finished")
+    } yield ()
+  }
+
+  def runRevokeExpiredCredits(): Fox[Unit] =
+    for {
+      orgas <- findAllOrganizationsWithExpiredCredits
+      failedTransactionsToRevoke <- Fox.foldLeft(orgas.iterator, List(): List[CreditTransaction]) {
+        case (failedTransactions, organizationId) =>
+          run(revokeExpiredCreditsForOrganization(organizationId).transactionally).map(failedTransactions ++ _)
+      }
+      _ = if (failedTransactionsToRevoke.nonEmpty) {
+        val failedTransactions = failedTransactionsToRevoke.map(transaction =>
+          s"Failed to revoke credits for transaction ${transaction._id} for organization ${transaction._organization}.")
+        slackNotificationService.warn("Failed to revoke some expired credits for organizations",
+                                      s"${failedTransactions.mkString("\n")}")
+      }
+    } yield ()
```

Comment:
I’m not sure I understand these Revoke functions. Maybe let’s talk about this in person.